solver/src/http_solver/settlement.rs

```diff
@@ -18,92 +21,172 @@ pub struct PreparedModel {
-        };
-        if executed_amount.is_zero() {
-            continue;
+// An intermediate representation between SettledBatchAuctionModel and Settlement useful for doing
+// the error checking up front and then working with a more convenient representation.
+struct IntermediateSettlement {
+    executed_limit_orders: Vec<ExecutedLimitOrder>,
+    executed_amms: Vec<ExecutedAmm>,
+    prices: HashMap<H160, U256>,
+}
+
+struct ExecutedLimitOrder {
+    order: LimitOrder,
+    executed_buy_amount: U256,
+    executed_sell_amount: U256,
+}
+
+impl ExecutedLimitOrder {
+    fn executed_amount(&self) -> U256 {
+        match self.order.kind {
+            OrderKind::Buy => self.executed_buy_amount,
+            OrderKind::Sell => self.executed_sell_amount,
+        }
+    }
+}
+
+struct ExecutedAmm {
+    order: AmmOrder,
+    input: (H160, U256),
+    output: (H160, U256),
+}
+
+impl IntermediateSettlement {
+    fn new(settled: SettledBatchAuctionModel, prepared: PreparedModel) -> Result<Self> {
+        let executed_limit_orders = new_orders(prepared.limit_orders, settled.orders)?;
+        let executed_amms = new_amms(prepared.amm_orders, settled.uniswaps)?;
+        let prices = new_prices(
+            &prepared.tokens,
+            executed_limit_orders.as_slice(),
+            executed_amms.as_slice(),
+            settled.prices,
+        )?;
+        Ok(Self {
+            executed_limit_orders,
+            executed_amms,
+            prices,
+        })
+    }
+
+    fn into_settlement(self) -> Settlement {
+        let mut settlement = Settlement::default();
+        for order in self.executed_limit_orders.iter() {
+            let (trade, interactions) = order
+                .order
+                .settlement_handling
+                .settle(order.executed_amount());
+            if let Some(trade) = trade {
+                settlement.trades.push(trade);
+            }
+            settlement.interactions.extend(interactions);
         }
-        let (trade, interactions) = order.settlement_handling.settle(executed_amount);
-        if let Some(trade) = trade {
-            settlement.trades.push(trade);
+        for amm in self.executed_amms.iter() {
+            let interactions = amm.order.settlement_handling.settle(amm.input, amm.output);
+            settlement.interactions.extend(interactions);
         }
-        settlement.interactions.extend(interactions);
+        settlement.clearing_prices = self.prices;
+        settlement
     }
-    Ok(())
 }
 
-fn set_amms(
-    model: &SettledBatchAuctionModel,
-    amms: &HashMap<String, AmmOrder>,
-    settlement: &mut Settlement,
-) -> Result<()> {
-    for (index, model) in model.uniswaps.iter() {
-        let amm = amms
-            .get(index.as_str())
-            .ok_or_else(|| anyhow!("invalid amm {}", index))?;
-        let (input, output) =
-            if model.balance_update1.is_positive() && model.balance_update2.is_negative() {
+fn new_orders(
```

Comment:
nit: I find this method name very vague. Maybe something like `match_prepared_and_settled_orders`?